src/profile-logic/transforms.js

```diff
@@ -688,6 +730,78 @@ export function collapseResource(
   });
 }
 
+export function collapseDirectRecursion(
+  thread: Thread,
+  funcToCollapse: IndexIntoFuncTable,
+  implementation: ImplementationFilter
+): Thread {
+  const { stackTable, frameTable, samples } = thread;
+  const oldStackToNewStack: Map<
+    IndexIntoStackTable | null,
+    IndexIntoStackTable | null
+  > = new Map();
+  oldStackToNewStack.set(null, null);
+  const nonRecursiveStacks = new Set([null]);
```

Comment:
Could you invert the meaning of this and call it `recursiveStacks`? I think that would make the checks below easier to understand. And I would expect to have fewer recursive stacks than non-recursive stacks on average, but maybe that isn't true.